Merged Keeney's changes

dkeeney opened the pull request:
Added Hunter for package manager
Configured CMake
Added SPRegion,
Added Unit tests
Removed Dimensions and LinkPolicy.
Lots of code cleanup.

dkeeney committed:
Merged Keeney's changes

Added Hunter for package manager
Configured CMake
Added SPRegion,
Added Unit tests
Removed Dimensions and LinkPolicy.
Lots of code cleanup.

dkeeney committed:
Set compiler to c++17
Fixed things that it broke.
Removed zlib from dependancies
Other cleanups.

chhenning commented:
Thanks @dkeeney! This looks much better! Can you quickly describe why we need the hunter package manager?

dkeeney commented:
Sure,
We need dependencies to be installed somehow.  They could be installed
independently but they most likely will be not be installed into 'standard'
or 'known' places.  They could be installed anywhere (more so on Windows
than on Linux). Therefore it takes a manual effort to tell our build
process where to find them.

What Hunter does is find the correct installation package in its database,
one that matches our environment.  It will download it to Hunter's known
location, build it if needed, and then make it available to CMake so that
our build can find them.  All without any manual intervention.  It will
also make sure that updates are automatically downloaded and installed when
they become available (unless a specific version is specified).

I did discover that the package versions must be registered with Hunter by
the package author's before they will make them available (security
thing).  The yaml-cpp package is old, V0.5.2 and the current version is
V0.6.2.  I don't think that is a problem however. The rest of the packages
are the latest.

Now I agree that some people might WANT to use independently installed
dependencies. So we should also probably include a build.bat that can be
edited with their locations for the dependencies rather than use Hunter.  I
have not built that.

Dave
======

On Mon, Jun 18, 2018 at 9:31 AM, chhenning <notifications@github.com> wrote:

> Thanks David! This looks much better! Can you quickly describe why we need
> the hunter package manager?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/chhenning/nupic.core/pull/4#issuecomment-398115065>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBa___rl53WEWPF6MWO7hQBVBqyXfcbks5t99XagaJpZM4Uq1ed>
> .
>

chhenning commented:
I'm not sure we want to install packages for our client. If cmake cannot find a package the script will just complain. My preference would be to ignore the package managers for now. This is highly charged issue and as far as I can see most open source projects just provide a cmake script without package manager dependencies. 

Do you agree?

dkeeney committed:
More code cleanup.
Got rid of NTA_xxx data types, replacing them with ones defined in types.hpp.
fixed some more compiler warnings.
Added some more tests for type conversion method.

dkeeney commented:
Ok, but CMake does not just search everywhere on your computer looking for
packages. It has to be told where to look. The Find_Package function does
not just look everywhere on your computer to find a package.

I have the script FindBoost.cmake which will look at the environment
variables BOOST_DIR or BOOST_ROOT which the user must have set up.  Once it
has that it can use the Find_Package function to do the rest.

The Python library can be found by running Python and asking it where the
library is.  Not so easy with other packages.

I will make a build.bat which will does not use the package manager and see
how well it works.

Dave
==========


On Mon, Jun 18, 2018 at 11:48 AM, chhenning <notifications@github.com>
wrote:

> I'm not sure we want to install packages for our client. If cmake cannot
> find a package the script will just complain. My preference would be to
> ignore the package managers for now. This is highly charged issue and as
> far as I can see most open source projects just provide a cmake script
> without package manager dependencies.
>
> Do you agree?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chhenning/nupic.core/pull/4#issuecomment-398156653>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBa_2gxcLA9oUMKDiQaiz9e4eXKZkuMks5t9_XxgaJpZM4Uq1ed>
> .
>

chhenning commented:
I thought nupic.core doesn't rely on python anymore? If it does such code should go into the nupic.python layer.

Yes, you need to supply the folders to the build script. Just like the example.bat that I provide in external/windows/ .
Btw, if you build all the dependencies with vcpkg all you need to supply is the cmake toolchain file. vcpkg keeps it updated with all the libraries installed. On my machine it's here:

D:\vcpkg\scripts\buildsystems\vcpkg.cmake

dkeeney commented:
Christian,
No, it does not need python for nupic.core.  But we will need it for the
Python interface when we do that package.

Ah, yes if you have your own package manager like vcpkg then you will know
where to look.  I assume that most people will not be using something like
that.

Lets see what I can come up with...this should be doable.

Dave
=======

On Tue, Jun 19, 2018 at 6:41 AM, chhenning <notifications@github.com> wrote:

> I thought nupic.core doesn't rely on python anymore? If it does such code
> should go into the nupic.python layer.
>
> Yes, you need to supply the folders to the build script. Just like the
> example.bat that I provide in external/windows/ .
> Btw, if you build all the dependencies with vcpkg all you need to supply
> is the cmake toolchain file. vcpkg keeps it updated with all the libraries
> installed. On my machine it's here:
>
> D:\vcpkg\scripts\buildsystems\vcpkg.cmake
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/chhenning/nupic.core/pull/4#issuecomment-398402961>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFBa_wgtWwcypoq9Q9TkS3j7YLpgD4fFks5t-P-lgaJpZM4Uq1ed>
> .
>

dkeeney committed:
New startup scripts. One for vcpkg and one for hunter.
Minor fixes to clean up warnings.
Removed need for ZLib.

dkeeney committed:
Another attempt to get both vcpkg to build.

dkeeney committed:
Startup can use either vcpkg or hunter package manager.

dkeeney committed:
Converted more boost:: functions to std:: functions.

dkeeney committed:
Finished coding/compiling BacktrakingTMCpp.cpp/hpp
NOTE: Debugging and unit tests not yet finished.

dkeeney committed:
Finished the backtrackingTM testing.

breznak commented on CMakeLists.txt:
```diff
@@ -329,6 +329,8 @@ set(types_files
 )
 
 set(utils_files
+    ${nupic_src}/utils/CSVReader.cpp
```
consider using https://github.com/ben-strasser/fast-cpp-csv-parser/blob/master/csv.h 
I've worked with it, it's include only, might be a better fit.

dkeeney committed:
Reimplemented MathsTest as part of unit test.
Removed the last of the boost functions.
Removed unused file GraphAlgorithms.hpp
Removed the boost dependancy.

dkeeney committed:
Made Yaml serialization a compile option.
By default it will do stream serialization.
This was an attempt to reduce dependancies
however, Yaml is still required for Region
creation parameters.

dkeeney committed:
Revert "Made Yaml serialization a compile option."

Decided this was a bad idea. We have to have Yaml for parameter parsing
anyway so we cannot remove it as a dependancy.

This reverts commit b5881a652a79433cf0a34d5c8bf1a81fdcbb24f5.

dkeeney committed:
Removed os.FStream.cpp and .hpp.
They were redundant to std::istream and std::ostream.

breznak commented on CMakeLists.txt:
```diff
@@ -290,8 +290,6 @@ set(os_files
     ${nupic_src}/os/Env.cpp
     ${nupic_src}/os/Env.hpp
-    ${nupic_src}/os/FStream.cpp
-    ${nupic_src}/os/FStream.hpp
```
Did you actually remove the files? I don't see it in the PR, but might be github just doesn't show it. 
Great work and nice commit (it'll be easy to merge).

dkeeney committed:
Addded type range chacking to BasicType conversions.

dkeeney committed:
Synchronized Serialization code with community.
